File: project_pos/lib/features/customers/services/customer_vehicle_service.dart
import 'package:project_pos/core/api/api_client.dart';
import 'package:project_pos/core/utils/app_logger.dart';

/// Sprint 10 — CustomerVehicle (parçacı sektör plaka takibi) HTTP servisi.
///
/// Backend endpoint kataloğu (Sprint 9):
///   GET    /customers/{id}/vehicles
///   GET    /customers/{id}/vehicles/search?q=
///   GET    /customers/{id}/vehicles/{vid}
///   POST   /customers/{id}/vehicles                 (idempotent)
///   PUT    /customers/{id}/vehicles/{vid}
///   DELETE /customers/{id}/vehicles/{vid}           (soft-delete)
class CustomerVehicleService {
  final ApiClient _apiClient;

  CustomerVehicleService(this._apiClient);

  static String _base(String customerId) =>
      'product/api/v1/customers/$customerId/vehicles';

  /// Müşterinin aktif plakaları.
  Future<List<Map<String, dynamic>>> listByCustomer(String customerId) async {
    try {
      final response = await _apiClient.get(_base(customerId));
      return List<Map<String, dynamic>>.from(response.data['data'] ?? []);
    } catch (e) {
      AppLogger.error('CustomerVehicle list hatası',
          tag: 'CustomerVehicleService', error: e);
      rethrow;
    }
  }

  /// Plaka prefix arama (autocomplete; backend normalize eder).
  Future<List<Map<String, dynamic>>> search(
      String customerId, String query) async {
    try {
      final response = await _apiClient.get(
        '${_base(customerId)}/search',
        queryParameters: {if (query.isNotEmpty) 'q': query},
      );
      return List<Map<String, dynamic>>.from(response.data['data'] ?? []);
    } catch (e) {
      AppLogger.error('CustomerVehicle search hatası',
          tag: 'CustomerVehicleService', error: e);
      rethrow;
    }
  }

  /// Yeni plaka ekleme — backend idempotent: aynı normalized varsa mevcut döner.
  Future<Map<String, dynamic>> create(
      String customerId, Map<String, dynamic> data) async {
    try {
      final response = await _apiClient.post(_base(customerId), data: data);
      return response.data['data'] as Map<String, dynamic>;
    } catch (e) {
      AppLogger.error('CustomerVehicle create hatası',
          tag: 'CustomerVehicleService', error: e);
      rethrow;
    }
  }

  Future<Map<String, dynamic>> update(
      String customerId, String vehicleId, Map<String, dynamic> data) async {
    try {
      final response =
          await _apiClient.put('${_base(customerId)}/$vehicleId', data: data);
      return response.data['data'] as Map<String, dynamic>;
    } catch (e) {
      AppLogger.error('CustomerVehicle update hatası',
          tag: 'CustomerVehicleService', error: e);
      rethrow;
    }
  }

  /// Soft-delete (isActive=false).
  Future<void> deactivate(String customerId, String vehicleId) async {
    try {
      await _apiClient.delete('${_base(customerId)}/$vehicleId');
    } catch (e) {
      AppLogger.error('CustomerVehicle deactivate hatası',
          tag: 'CustomerVehicleService', error: e);
      rethrow;
    }
  }
}

File: pos-product-manager/src/main/java/com/sedcore/customer/service/CustomerVehicleService.java
package com.sedcore.customer.service;

import com.sedcore.customer.entity.CustomerVehicle;
import com.sedcore.customer.model.CustomerVehicleDto;
import com.sedcore.customer.model.CustomerVehicleResponse;
import com.sedcore.customer.model.VehicleSearchResponse;

import java.util.List;
import java.util.Optional;

/**
 * Sprint 9 — CustomerVehicle service interface.
 *
 * <p>@Service annotated impl AOP CompanyHibernateFilterActivator pointcut'ına
 * uyar (com.sedcore..service..*) → tenant filter aktif olur.
 */
public interface CustomerVehicleService {

    /** Müşterinin aktif plakaları (UI dropdown için). */
    List<CustomerVehicleResponse> listByCustomer(String customerId);

    /** Plaka prefix arama (autocomplete). */
    List<CustomerVehicleResponse> searchByCustomer(String customerId, String q);

    /** Tek kayıt. */
    Optional<CustomerVehicleResponse> findById(String id);

    /** Entity getter — Sale.createSale içinde FK kullanmak için. */
    CustomerVehicle getEntity(String id);

    /** Yeni plaka kaydı (idempotent: aynı normalized varsa mevcut döner). */
    CustomerVehicleResponse create(String customerId, CustomerVehicleDto dto);

    CustomerVehicleResponse update(String id, CustomerVehicleDto dto);

    /** Soft-delete (isActive=false). */
    CustomerVehicleResponse deactivate(String id);

    /**
     * Sprint 11e — Tenant-wide plaka prefix arama (customerId yok).
     * Her sonuç müşteri ismi + açık satış count + açık tutar içerir.
     * Limit varsayılan 20.
     */
    List<VehicleSearchResponse> searchAcrossTenant(String q, int limit);
}
